data/scenarios/Challenges/_friend/cat.sw

```diff
@@ -0,0 +1,70 @@
+def forever : cmd unit -> cmd unit = \c. c ; forever c end
+
+def repeat : int -> cmd unit -> cmd unit =
+  \n. \c. if (n == 0) {} {c ; repeat (n-1) c}
+end
+
+def elif = \t. \then. \else. {if t then else} end
+def else = \t. t end
+
+def abs = \n. if (n < 0) {-n} {n} end
+
+def randdir : cmd dir =
+  d <- random 4;
+  return (
+    if (d == 0) {north}
+    $ elif (d == 1) {east}
+    $ elif (d == 2) {south}
+    $ else {west}
+  )
+end
+
+def wander =
+  n <- random (16*4);
+  wait (16*2 + n);
+  d <- randdir;
+  turn d;
+  dist <- random 2;
+  try {repeat dist move} {};
+  r <- random 5;
+  if (r == 0) { say "meow" } {}
+end;
+
+def disappointed = \cat.
+  say "meow??";
+  n <- count "fish";
+  cat n
+end
+
+def follow : (int -> cmd unit) -> actor -> cmd unit = \cat. \r.
+  rLoc <- as r {whereami};
+  myLoc <- whereami;
+  let dx = fst rLoc - fst myLoc in
+  let dy = snd rLoc - snd myLoc in
+  if (abs dx > abs dy)
+  { if (dx < 0) {turn west} {turn east} }
+  { if (dy < 0) {turn south} {turn north} };
+  if (abs dx != 0 || abs dy != 0) {try { move } { disappointed cat }} {};
+  wait 4;
+  follow cat r
+end
+
+def love = \cat.
+  say "purr";
+  fishGiver <- meet;
+  case fishGiver
+    (\_. say "meow??"; n <- count "fish"; cat n)
+    (\r. follow cat r)
+end
+
+def cat = \fishCount.
+  n <- count "fish";
+  if (n > fishCount)
+  { say "yum!";
+    if (n >= 3) { love cat } { cat n }
```

Comment:
Here's my experience report for solving this :)

* Was initially concerned that I'd have to write a program to locate the cat at some random displacement, but soon figured that, upon starting over, just making a mad dash straight to the cat is reproducible.
* Unfortunately I was tainted by seeing the number 3 in the code review.  I tried anyway just feeding one fish before dashing back to my base and got the "meow??".  I think that's a good enough hint.
* I again immediately dashed away after feeding 3 fish and got a `meow??`.  I figured maybe the cat needed some "bonding time" or maybe I didn't stick around long enough for them see who left the fish :).   So I retried a few times with increasing `wait` times until it worked.

Reply:
Thanks for the experience report!  Yeah, the needed "bonding time" is because the cat repeatedly picks a random amount of time to wait before it (1) tries to move and (2) notices if it has more fish.  I could probably try to change it so it notices fish immediately.  Do you think that would be better, or it's OK the way it is?

Reply:
I think it would be better to notice the fish immediately.

Reply:
If there is still an issue with getting the fish to be immediately acknowledged, I think a more detailed hint---perhaps a Garfield-style internal monologue, like "Meow?? (who gave that fish?)"---would help.

Reply:
OK, I changed it so the cat is constantly polling to see if it has more fish than before.  This means you usually don't have to wait for the cat to follow you, although in my testing it is possible, if you give the fish and then *immediately* run away, to get far enough away before it runs `meet` than you get the `meow??`.  But other than that small race condition it starts following you right away.